mutiny-core/src/federation.rs

```diff
@@ -0,0 +1,612 @@
+
+        log_debug!(logger, "Built fedimint client");
+        Ok(FederationClient {
+            uuid,
+            federation_index: federation_index.clone(),
+            federation_code,
+            fedimint_client,
+            stopped_components,
+            storage,
+            network,
+            logger,
+            stop,
+        })
+    }
+
+    pub(crate) fn federation_index(&self) -> FederationIndex {
+        FederationIndex {
+            child_index: self.federation_index.child_index,
+            federation_code: self.federation_code.clone(),
+            archived: self.federation_index.archived,
+        }
+    }
+
+    pub(crate) async fn get_invoice(&self, amount: u64) -> Result<MutinyInvoice, MutinyError> {
+        let lightning_module = self
+            .fedimint_client
+            .get_first_module::<LightningClientModule>();
+        let (_id, invoice) = lightning_module
+            .create_bolt11_invoice(Amount::from_sats(amount), String::new(), None, ())
+            .await?;
+        Ok(invoice.into())
+    }
+
+    /// Get the balance of this federation client in sats
+    pub(crate) async fn get_balance(&self) -> Result<u64, MutinyError> {
+        Ok(self.fedimint_client.get_balance().await.msats / 1_000)
+    }
+
+    pub async fn get_activity(&self) -> Result<Vec<ActivityItem>, MutinyError> {
+        let operations = self
+            .fedimint_client
+            .operation_log()
+            .list_operations(100, None)
+            .await;
+
+        let lightning_module = Arc::new(
+            self.fedimint_client
+                .get_first_module::<LightningClientModule>(),
+        );
+
+        let activity_stream = stream::iter(operations.into_iter());
+
+        let activity_items = activity_stream
+            .then(move |(key, entry)| {
+                let logger = self.logger.clone();
+                let lightning_module = Arc::clone(&lightning_module);
+                async move {
+                    if entry.operation_module_kind() == "ln" {
+                        Some(
+                            handle_ln_meta(logger, &entry, key.operation_id, &lightning_module)
+                                .await,
+                        )
+                    } else {
+                        log_warn!(
+                            logger,
+                            "Unsupported module: {}",
+                            entry.operation_module_kind()
+                        );
+                        None
+                    }
+                }
+            })
+            .filter_map(|item| async move { item })
+            .collect::<Vec<ActivityItem>>()
+            .await;
+
+        Ok(activity_items)
+    }
+
+    pub async fn get_invoice_by_hash(
+        &self,
+        hash: &sha256::Hash,
+    ) -> Result<MutinyInvoice, MutinyError> {
+        let operations = self
+            .fedimint_client
+            .operation_log()
+            .list_operations(100, None)
```

Comment:
i think this only looks back at the last hundred, we'll need to add handling if you've done over 100 txs